src/main/jbake/content/blog/2017/06/23/opentracing-appmetrics.adoc

```diff
@@ -0,0 +1,270 @@
+instrumentation within the two services:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-grafana-dashboard.png[caption="Figure 4: ", title="Grafana dashboard"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-grafana-dashboard.png[caption="Figure 4: ", title="Grafana dashboard"]
+endif::[]
+
+This dashboard includes three graphs, the first showing the span count, the second showing the average duration
+of the spans, and third showing the ratio between successful and erronous spans.
+
+The metrics reported by Prometheus are based on a range of labels - a metric exists for each unique combination
+of those labels.
+
+The standard labels included with the OpenTracing java-metrics project are: `operation`, `span.kind` and `error`.
+
+With this particular example, we also included the `transaction` label.
+
+However when the services are deployed to Kubernetes, the following additional labels are included for free: `pod`,
+`instance`, `service`, `job` and `namespace`.
+
+In our example Prometheus queries, we have ignored most of the Kubernetes added labels (except `service`) so that
+the metrics are aggregated across the specific pods, namespaces, etc. However, having these labels available means
+it is possible to segment the metrics in whatever way is required to analyse the data.
+
+When using the `java-metrics` project outside of Kubernetes, it is still possible to include the `service`
+label, however this would be configured when setting up the tracer.
+
+We can also filter the data, to focus on specific areas of interest:
+
+ifndef::env-github[]
+image::/img/blog/2017/2017-06-23-grafana-txn-service.png[caption="Figure 5: ", title="Customized Grafana graph focusing on metrics for transaction 'sell' and service 'service2'"]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/2017-06-23-grafana-txn-service.png[caption="Figure 5: ", title="Customized Grafana graph focusing on metrics for transaction 'sell' and service 'service2'"]
+endif::[]
+
+In this image we have filtered the metrics based on the `transaction='sell'` and `service='service2'`. This is
+where using the metric label based on the baggage item `transaction` can be useful, to understand the
+usage of a particular shared service by a business transaction. With further work it would be possible
+to show the distribution of requests for a service across the various business transactions.
+
+
+== Conclusion
+
+This article has shown how a service can be instrumented once (using OpenTracing) and generate both
+tracing and application metrics.
+
+When deployed to a Kubernetes environment, the metrics also benefit from additional labels automatically
+add to the metrics describing the service, pod, namespace, etc, that can make it easy to isolate
+specific metrics of interest, or view high level aggregated metrics to gain an overview of your
+applications performance.
+
```

Comment:
The last sentence needs to be reworded for clarity. I’m not sure if you mean “additional labels automatically added to the metrics” or if I'm misreading this sentence entirely.